openmmapi/src/VirtualSite.cpp

```diff
@@ -184,3 +184,29 @@ Vec3 LocalCoordinatesSite::getYWeights() const {
 const Vec3& LocalCoordinatesSite::getLocalPosition() const {
     return localPosition;
 }
+
+SymmetrySite::SymmetrySite(int particle, const Vec3& Rx, const Vec3& Ry, const Vec3& Rz, const Vec3& v, bool useBoxVectors) :
+        Rx(Rx), Ry(Ry), Rz(Rz), v(v), useBoxVectors(useBoxVectors) {
+    setParticles({particle});
+    Vec3 R[] = {Rx, Ry, Rz};
+    for (int i = 0; i < 3; i++)
```

Comment:
This sanity check will fail some valid symmetry operators in fractional coordinates. For example, the space group P 63 includes the following fractional-coordinate rotation matrices, each of which gives a dot product of 1.0 between at least one pair of axes - but when combined with the box vectors they yield rigid-body rotations in Cartesian space.

```
[array([[ 0.,  1.,  0.],
        [-1.,  1.,  0.],
        [ 0.,  0.,  1.]]),
 array([[ 1., -1.,  0.],
        [ 1.,  0.,  0.],
        [ 0.,  0.,  1.]]),
 array([[-1.,  1.,  0.],
        [-1.,  0.,  0.],
        [ 0.,  0.,  1.]]),
 array([[ 0., -1.,  0.],
        [ 1., -1.,  0.],
        [ 0.,  0.,  1.]])]
]
```

Reply:
That will complicate things.  We assume it's an orthogonal matrix, so its transpose is the inverse.  We use that in applying the forces.  Presumably if we explicitly invert the matrix, the same approach could be applied to any non-singular matrix?  I'm trying to think what weird situations could come up when dealing with non-orthogonal matrices.

Reply:
Hi Peter, You could leave the sanity check, but only if the "useBoxVectors" is false?

A 2nd option is to remove the "useBoxVectors" flag, but have the API require supplying a "total" transformation operator:
T = toCart * Rot * toFrac 
where toCart, Rot, toFrac are 3x3 matrices chained together (Rot is the fractional symmetry operator rotation matrix).
To rotate forces from the symmetry site back to the "real" site the transpose of T can be used as you suggest above. A big drawback in this case is that toCart & toFrac matrices change with box vectors (e.g. for NPT).

A 3rd option could be to supply box vectors, generate the toCart / toFrac matrices, and then test the orthogonality of T. 

I hope this helps.

Reply:
>  T = toCart * Rot * toFrac

That only works if there's no translation.  To be able to compose the whole transformation like that, we would need to represent it as a 4x4 matrix and work in homogeneous coordinates.  It also wouldn't allow you to change the box vectors.  Though possibly that's a reasonable restriction for this feature?

Reply:
If you were to allow the box vectors to change, it would probably make most sense to enforce the edge length and angle rules of the underlying [Bravais lattice](https://en.wikipedia.org/wiki/Bravais_lattice#In_3_dimensions) for the corresponding space group (which, of course, would require the user to specify the space group)... although I suppose these rules would naturally emerge from the simulations anyway since they're the essentially the conditions under which transforms in fractional coordinates correspond to rigid-body (or mirror) transforms in real space. One would expect that any change in the box that leads to stretching or compression of the symmetry copies would be very strongly resisted.

I'd lean towards keeping the current system. For the up-front sanity check, you could combine with the current box vectors to generate the Cartesian transform and then see if _that_ passes the existing check. During the simulation, the advantage of working in fractional coordinates is that the transformation matrix never changes so you only need to do the inversion once and store the result.

Reply:
There aren't any current box vectors at this point in the code.  It's just a VirtualSite object that hasn't even been added to a System yet.  We could delay all the validation to the Context constructor, but that will fail if the user hasn't bothered to set the default box vectors and plans to set the correct box vectors after creating the Context.

How about keeping this check only when working in Cartesian coordinates?  When using fractional coordinates, it will invert the matrix and only throw an exception if it's singular.  It's up to the user to pick a matrix that will work correctly with their box shape.

Reply:
As you note @peastman the above only considered the rotational part of the transform. You might want to keep the rotational and translational transforms as separate 3x3 matrices (instead of using 4x4 matrices), since only the transpose of the rotational part is needed to rotate forces.
Rot_Cart = toCart * Sym.Rot * toFrac
Trans_Cart = toCart * [I + Sym.Trans] * toFrac
The Sym.Trans matrix has the SymOp translation in fractional coordinates on the diagonal and "I" is the Identity matrix. The constructor would have two 3x3 matrices passed in (Rot_Cart & Trans_Cart). Advantages are the ability to check the orthogonality of both operators, the VirtualSite doesn't need to know the box vectors, you're always working in Cartesian coordinates, and the rotation of forces is just the transpose of Rot_Cart. A drawback is the box vectors need to remain fixed, or alternatively arrange an O(N) update of the VirtualSite instances with new transforms if/when box vectors change. As @tristanic points out, it would be good to enforce lattice system constraints some place.

Reply:
I don't understand.  How do you represent a translation as a 3x3 matrix?

Reply:
@peastman  -- my apologies. The Trans_Cart would need to be a 4x4 Identity matrix plus the Sym.Trans as the first 3 entries in the last column. The "toFrac" would need to produce a 4 element vector [Fa, Fb, Fc, 1]. Same idea overall idea though.

Reply:
Ok, so you're using homogeneous coordinates.  I guess the question is whether we should allow specifying transformations in fractional coordinates, or whether we should require Cartesian transformations.  The latter is more difficult for the user since they have to compute the compound transformations themself, and it requires the box vectors to be baked into the transformation.  But it does allow us to do the sanity check on the rotation matrix.

Reply:
@peastman  I defer to your judgement. To make it easier on the User, a method to generate the compound transformation(s) could be supplied? One other thing to consider before I forget is you may want to check for "special positions" when generating coordinates for symmetry sites. PDB files often have water or ions at special positions, but at fractional occupancy (e.g., half of the water belongs to the asymmetric unit, and half belongs to the symmetry mate). Application of the SymOp leaves the coordinates unchanged (or nearly so), and then you've got an infinite energy / force if the particle pair interact.

Reply:
> DB files often have water or ions at special positions, but at fractional occupancy (e.g., half of the water belongs to the asymmetric unit, and half belongs to the symmetry mate). Application of the SymOp leaves the coordinates unchanged (or nearly so), and then you've got an infinite energy / force if the particle pair interact.

We had a [long discussion](https://github.com/openmm/openmm/issues/1437#issuecomment-684649081) on the subject of special positions and waters a few years back. The upshot is that I think applying the crystallographic symmetry to solvent molecules in an equilibrium simulation is pretty much a non-starter - if you want explicit solvent then I think the better solution will be some hybrid where the protein is symmetry-constrained while the unit cell is flooded with unconstrained water. Ultimately I'm not sure there's one true way to handle interactions around special positions - the strategy will vary depending on what you're trying to achieve.